### ISSUE TITLE
Freeze when using slow storage

### ISSUE THREAD

**Issue by finwo** [syncthing/syncthing#6503]
When using slow/unstable storage, syncthing becomes barely usable. While this is actually an issue with the filesystem or storage provider, syncthing still produces unexpected behavior.

## What happened

On every filesystem action (assumed), syncthing freezes for the duration of the operation (again, assumed). During the freeze, the web gui does not respond (refreshes and api calls hang).

## What you expected to happen

Slow transfers but the web gui still operational

## Steps to reproduce the problem

The gcsfuse adapter for google cloud buckets is slow, therefore I'll advise that to reproduce the problem.

* Create a google cloud account
* Use console.cloud.google.com to create a bucket
* Create a vm (compute engine), using debian or ubuntu
  * To save costs, select a micro machine
  * Under permissions, set storage to full
  * Enable http access
* Mount the bucket in the vm `gcsfuse --foreground <bucket> <mountpoint>`
* Start syncthing
* Stop syncthing
* Change `~/.config/syncthing/config.xml` to host the gui on `0.0.0.0:8080`
* Start syncthing again
  * Set username/password from the web gui
* Create a shared folder on the bucket's mountpoint
* Share the folder with another machine
* Add some files to the shared folder
* Observe the hosted one

Logs:

```
# systemctl status syncthing
● syncthing.service - NZBGet Service
   Loaded: loaded (/etc/systemd/system/syncthing.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2020-04-06 08:30:40 UTC; 7min ago
 Main PID: 724 (syncthing)
    Tasks: 19 (limit: 681)
   CGroup: /system.slice/syncthing.service
           ├─724 /usr/bin/syncthing
           └─795 /usr/bin/syncthing

Apr 06 08:31:09 syncthing syncthing[724]: [5MZGP] 08:31:08 INFO: Established secure connection to POYLTB2-*** at 10.164.0.3:44952-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:31:09 syncthing syncthing[724]: [5MZGP] 08:31:09 INFO: Device POYLTB2-*** client is "syncthing v1.4.0" named "MRD-LX1" at 10.164.0.3:44952-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:31:09 syncthing syncthing[724]: [5MZGP] 08:31:09 INFO: Connection to POYLTB2-*** at 10.164.0.3:44952-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256 closed: reading length: read tcp 10.164.0.3:44952->***:18579: read: connection reset by peer
Apr 06 08:31:55 syncthing syncthing[724]: [5MZGP] 08:31:55 INFO: Established secure connection to POYLTB2-*** at 10.164.0.3:45492-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:31:56 syncthing syncthing[724]: [5MZGP] 08:31:55 INFO: Device POYLTB2-*** client is "syncthing v1.4.0" named "MRD-LX1" at 10.164.0.3:45492-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:31:56 syncthing syncthing[724]: [5MZGP] 08:31:56 INFO: Connection to POYLTB2-*** at 10.164.0.3:45492-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256 closed: reading length: read tcp 10.164.0.3:45492->***:18579: read: connection reset by peer
Apr 06 08:31:57 syncthing syncthing[724]: [5MZGP] 08:31:57 INFO: Joined relay relay://***:22067
Apr 06 08:32:52 syncthing syncthing[724]: [5MZGP] 08:32:52 INFO: Established secure connection to POYLTB2-*** at 10.164.0.3:45532-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:32:52 syncthing syncthing[724]: [5MZGP] 08:32:52 INFO: Device POYLTB2-*** client is "syncthing v1.4.0" named "MRD-LX1" at 10.164.0.3:45532-***:18579/tcp-client/TLS1.3-TLS_CHACHA20_POLY1305_SHA256
Apr 06 08:32:52 syncthing syncthing[724]: [5MZGP] 08:32:52 INFO: Unexpected folder "Camera" (mrd-lx1_***-photos) sent from device "POYLTB2-***"; ensure that the folder exists and that this device is selected under "Share With" in the folder configuration.
```

### Version Information

Syncthing Version: v1.4.0
OS Version: Ubuntu 18.04
Browser Version: Chromium Version 81.0.4044.62 (Developer Build) (64-bit)

**Comment by calmh** [comment]
Just to be clear here, it's only a shared folder that is on crap storage, yeah? If it's the database/config I would say the observed behavior is expected. If it's just a folder, I'd say it's unexpected except for the few API calls that access the folder in question (scans, reading ignore patterns, ...).

**Comment by imsodin** [comment]
I think the GUI isn't handling config changes that take long particularly well. And if you do changes like you describe while e.g. a folder scan is going on, it can take a long time for the config response to come in: The scan needs to be canceled first, and if a single operation during a scan can take a good while, then you have the lag you see.

**Comment by finwo** [comment]
@calmh The syncthing service runs as an unprivileged user & it's config.xml is in `~/.config/syncthing`. Therefor I assume the database/config you're talking about is not on the crap storage.

Keeping the network tab open, the event calls seem to continue so the freeze lies in the amount of network slots a tab as in the browser.

What I'd expect on crap storage, is the task being inserted in a queue of sorts & the request still finishing within reasonable time (and removing notifications like a folder being shared). The actual adding of the folder would then happen somewhere in the future.

**Comment by calmh** [comment]
So what are the things that are blocking precisely?

**Comment by finwo** [comment]
Adding a new share in the configuration described in the original.

The /rest/system/config on clicking the "Add" button of a new folder (or even the "share" when coming for a known folder) takes longer than I'm willing to wait (up to 30 minutes).

**Comment by calmh** [comment]
Right, OK, I didn't understand that in the initial post, sorry. Yeah, what @imsodin says; the config post will return when the config has been activated, which then requires waiting for some things to complete or abort. What is the other folder doing though? I think we have cancel points sprinkled fairly frequently.

**Comment by finwo** [comment]
Here's the current state of the machine.

![Screenshot 2020-04-06 at 135424](https://user-images.githubusercontent.com/6330094/78556675-af5a6180-780f-11ea-83e1-2729a46e0538.png)

Even if it's busy scanning, I'd expect a config change to be fairly quick. If the transfers are slow (e.g. the bulk of work), that's fine because it'll happen in the background.

**Comment by imsodin** [comment]
That's just not how the config system works currently. Not saying it is ideal how it is, but it would take rather large changes to accommodate "queuing" config changes. I can't remember an existing issue that covers this point, so imo this can stay for that.

For the question why it takes so long to abort the scan, such that the config change can go through, please open a thread on https://forum.syncthing.net/ so we can debug. If it turns out there's a bug/improvement there, we can open a issue on github specific to that.

**Comment by calmh** [comment]
I'd like to see a SIGQUIT trace to see what's going on when we're waiting for minute and minutes. 
Neither of the folders in the screenshot is doing any file i/o afaik ("preparing to sync" is "traversing the db", no?).

But apart from that this seems as-desired to me. We want config commits to be atomic - when the API call returns, the config has been applied. The alternative would be confusing and more or less require locking the GUI somehow anyway, even if with a spinner or progress bar or whatever. Having the GUI just continue on with the config *not* being changed as requested would be odd, it would seem just like nothing happened.

**Comment by finwo** [comment]
Right now there's no visual feedback anything happened. The desired behavior (imo) would be any sync activity not blocking config changes. Right now it's as if the config update and the disk activity runs in the same thread, and the disk activity using blocking IO.

**Without knowledge of the codebase**, just observing the gui, I'd try to move the config update or disk activity handler to a separate thread so the other is not blocked without feedback.

**Comment by finwo** [comment]
@calmh Just sent a SIGQUIT to all syncthing processes using htop while watching the log, none of the processes responded with a line in log.

How would I go about generating the trace you requested?

**Comment by AudriusButkevicius** [comment]
Sync activity is driven by the config, so a config change needs to rewire the sync process, which means in needs to terminate the existing one.

I feel this should be moved to the forum, as it's turning into a debugging session rather than a clearly quantified issue of which part of syncthing needs be fixed. Sure, the behaviour is undesired, but until we can nail down what exact part needs fix and whats the cause is, I don't think this belongs here.

**Comment by finwo** [comment]
For debugging and discussions: https://forum.syncthing.net/t/config-change-in-ui-taking-a-long-time-when-scanning-on-slow-storage/14764
